### PR TITLE
Add online since (runtime) label to stream cards + Stream View

### PR DESCRIPTION
**Pull request by Nickwasused**
# To do

- [x] Fix stuck stream time

This is fixing #36 

# The changes have been tested on:

- Android Virtual Pixel XL, Android 11.0
- Xiaomi Redmi Note 7, Android 11.0
- Samsung Galaxy Tab SM-T510, Android 10

I have tested every Theme of the App.

# Here is a Preview:
StreamCards:
Images:
[image]
[image]

## StreamView:
Images:
[image]
[image]

## Video:
[youtube](https://www.youtube.com/watch?v=kS9yi4Hy8oQ)

[youtube](https://www.youtube.com/watch?v=gDS1xQFnbBA)

### CODE REVIEW

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java [hunk 1]
```diff
@@ -376,6 +386,74 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
...
+                } catch (ParseException e) {
+                    e.printStackTrace();
+                }
+
+                int startminute = date.getMinutes();
+                int starthour = date.getHours();
+
+                // https://stackoverflow.com/questions/46774195/countdown-reverse-counter
+                // this is a reverse countdowntimer
+                new CountDownTimer(999999999, 1000) {
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java:407]: Couldn't this code be called from progressRunnable? That runs once a second, which is frequent enough. This reverse CountDownTimer approach seems a bit janky.

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java [hunk 2]
```diff
@@ -376,6 +386,74 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
...
+                                    mRuntime.setText("" + hour + ":0" + min + ":" + sec);
+                                }
+                            } else {
+                                if (hour < 10) {
+                                    mRuntime.setText("0" + hour + ":" + min + ":" + sec);
+                                } else {
+                                    mRuntime.setText("" + hour + ":" + min + ":" + sec);
+                                }
+                            }
+                        }
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java:449]: These if statements can be simplified:
```suggestion
                        mRuntime.setText(String.format("%-2s:%-2s:%-2s", hour, min, sec).replace(' ', '0'));
```

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java [hunk 3]
```diff
@@ -376,6 +386,74 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
             View mTimeController = mRootView.findViewById(R.id.time_controller);
             mTimeController.setVisibility(View.INVISIBLE);
 
+            if (!settings.getStreamPlayerRuntime()) {
+                mRuntimeWrapper.setVisibility(View.GONE);
+            } else {
+                Log.d("Runtime String", args.getString(getString(R.string.stream_fragment_runtime)));
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java:392]: Unnecessary logging.

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java [hunk 4]
```diff
@@ -376,6 +386,74 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
...
+                                } else {
+                                    mRuntime.setText("" + hour + ":" + min + ":" + sec);
+                                }
+                            }
+                        }
+                    }
+
+                    public void onFinish() {}
+                }.start();
+                //mRuntime.setText(String.valueOf(args.getString(getString(R.string.stream_fragment_runtime))));
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java:454]: I don't think this comment needs to be included.

**Review by samfundev — CHANGES_REQUESTED** [review]
Appreciate the PR! I found a few things to improve.

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java (hunk 4 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java]: Removed in https://github.com/twireapp/Twire/pull/211/commits/a190dbdb595d0bf8fdd47432b615b0556ef30c70

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java (hunk 3 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java]: Removed in https://github.com/twireapp/Twire/pull/211/commits/a190dbdb595d0bf8fdd47432b615b0556ef30c70

File: app/src/main/java/com/perflyst/twire/service/Settings.java [hunk 5]
```diff
@@ -655,6 +661,12 @@ public void setStreamPlayerShowViewerCount(boolean showViewCount) {
         editor.commit();
     }
 
+    public void setStreamRuntime(boolean showRuntime) {
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/service/Settings.java]: Changed in https://github.com/twireapp/Twire/pull/211/commits/a190dbdb595d0bf8fdd47432b615b0556ef30c70

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -72,6 +73,7 @@
 
     <string name="stream_fullscreen">Fullscreen</string>
     <string name="stream_viewers">Viewers</string>
+    <string name="stream_runtime"></string>
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/res/values/strings.xml]: The String has been removed in https://github.com/twireapp/Twire/pull/211/commits/a190dbdb595d0bf8fdd47432b615b0556ef30c70

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -275,6 +279,7 @@
     <!-- Changelog -->
     <string name="changelog_title">Changelog</string>
     <string name="changelog_show_after_update">Show changelog after an update</string>
+    <string name="player_show_runtime">Show the Stream Runtime</string>
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/res/values/strings.xml]: Moved to correct location in https://github.com/twireapp/Twire/pull/211/commits/a190dbdb595d0bf8fdd47432b615b0556ef30c70

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java (hunk 1 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java]: I have added its own runnable that runs every second in https://github.com/twireapp/Twire/pull/211/commits/bafd1f03537c69aa5f7ba141d6d3d7a4fe953569

File: app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java (hunk 2 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/fragments/StreamFragment.java]: I have added a cleaner way of calculating the Date in  therefore this is not needed anymore: https://github.com/twireapp/Twire/pull/211/commits/bafd1f03537c69aa5f7ba141d6d3d7a4fe953569

**Review by samfundev — APPROVED** [review]

**Comment by samfundev** [comment]
Everything looks good, merging. Many thanks!

**Pull request merged**